felix/fv/pre_dnat_test.go

```diff
@@ -200,6 +206,87 @@ var _ = infrastructure.DatastoreDescribe("pre-dnat with initialized Felix, 2 wor
 				})
 
 				Context("with pre-DNAT policy to open pinhole to 8055", func() {
+					// Exec iptables-save, or iptables-save -t [table], if table is not "".
+					dumpIPTables := func(table string) []string {
+						execArgs := []string{"iptables-save"}
+						if table != "" {
+							execArgs = append(execArgs, "-t", table)
+						}
+						out, err := tc.Felixes[0].ExecOutput(execArgs...)
+						Expect(err).NotTo(HaveOccurred())
+						lines := strings.Split(out, "\n")
+						Expect(len(lines)).NotTo(Equal(1), "couldn't split on newline")
+						var grep []string
+						for _, line := range lines {
+							if strings.Contains(line, "cali-") && strings.Contains(line, "-A ") {
+								grep = append(grep, line)
+							}
+						}
+						return grep
+					}
+					dumpMangleTable := func() []string { return dumpIPTables("mangle") }
+					dumpFilterTable := func() []string { return dumpIPTables("filter") }
+
+					// Returns nil when the value of f() hasn't changed for the duration settledPeriod.
+					// Returns error if f() never settles up to the timeout.
+					// Immediately returns error if timeout is lower than settledPeriod.
+					waitToSettle := func(f func() int, settledPeriod, timeout time.Duration) error {
+						if timeout < settledPeriod {
+							return errors.New("Programmer error: timeout is less than settledPeriod")
+						}
+						var last int
+						var settledTimer, timeoutTimer *time.Timer
+						timeoutTimer = time.NewTimer(timeout)
+						for {
+							cur := f()
+
+							if settledTimer != nil {
+								if last != cur {
+									if !settledTimer.Stop() {
+										<-settledTimer.C
+									}
+									settledTimer = nil
+								} else {
+									select {
+									case <-settledTimer.C:
+										// We've been settled long enough, return
+										return nil
+									default:
+										// Still settled, but not for long enough. Carry on...
+									}
+								}
+							} else {
+								if last == cur {
+									settledTimer = time.NewTimer(settledPeriod)
+								}
+							}
+							select {
+							case <-timeoutTimer.C:
+								return errors.New("Timed out waiting for IPTables to settle.")
+							default:
+								last = cur
+								time.Sleep(1 * time.Second)
```

Comment:
I typically reach for 100ms for a sleep loop.  We don't want to tight loop but probing 1x per second isn't normally a big deal.